src/AutoSubscriptions.ts

```diff
@@ -167,17 +173,27 @@ function _tryFinally<TResult>(tryFunc: () => TResult, finallyFunc: Function): TR
     }
 }
 
+function instanceTargetToInstanceTargetWithMetadata(instanceTarget: InstanceTarget, methodName?: string): InstanceTargetWithMetadata {
+    // Upcast here and make sure property exists
+    const newTarget = instanceTarget as InstanceTargetWithMetadata;
+    newTarget.__resubMetadata = newTarget.__resubMetadata || {};
+    if (methodName) {
```

Comment:
I think the "method" part should go into a separate method, like `resubMetadataToMethodMetadata` (or `instanceTargetWithMetadataToMethodMetadata`). Having this as an optional parameter is not great. Also, returning the method-metadata object is far safer than setting `__resubMetadata[methodName]` as a side effect.

It will lead to cleaner code as well, since `methodMetadata.hasAutoSubscribeDecorator` is easier to read than `target.__resubMetadata[methodName].hasAutoSubscribeDecorator`.